File: QTS/3° Bim/app-para-teste/web/app/models/Livro.php
<?php
// Model Livro
// ATENÇÃO: código intencionalmente com falhas para fins de estudo.

require_once __DIR__ . '/../core/Database.php';

class Livro
{
    private $db;

    public function __construct()
    {
        $this->db = Database::getInstance();
    }

    public function todos()
    {
        $resultado = $this->db->query("SELECT * FROM livros");
        return $resultado->fetchAll(PDO::FETCH_ASSOC);
    }

    // Busca livros por termo (pesquisa)
    public function buscar($termo)
    {
        // BUG DE SEGURANÇA + PERFORMANCE: SQL injection e LIKE sem escape, varredura total.
        $sql = "SELECT * FROM livros WHERE titulo LIKE '%$termo%' OR autor LIKE '%$termo%'";
        $resultado = $this->db->query($sql);
        if ($resultado === false) {
            return [];
        }
        return $resultado->fetchAll(PDO::FETCH_ASSOC);
    }

    public function porId($id)
    {
        $sql = "SELECT * FROM livros WHERE id = " . $id;
        $resultado = $this->db->query($sql);
        if ($resultado === false) {
            return null;
        }
        return $resultado->fetch(PDO::FETCH_ASSOC);
    }

    public function criar($titulo, $autor, $ano, $categoria, $quantidade, $isbn)
    {
        // BUG: não valida se quantidade é negativa antes do insert
        $sql = "INSERT INTO livros (titulo, autor, ano, categoria, quantidade, isbn)
                VALUES ('$titulo', '$autor', '$ano', '$categoria', $quantidade, '$isbn')";
        $this->db->exec($sql);
        return $this->db->getPdo()->lastInsertId();
    }

    public function atualizar($id, $titulo, $autor, $ano, $categoria, $quantidade, $isbn)
    {
        $sql = "UPDATE livros SET
                    titulo = '$titulo',
                    autor = '$autor',
                    ano = '$ano',
                    categoria = '$categoria',
                    quantidade = $quantidade,
                    isbn = '$isbn'
                WHERE id = $id";
        return $this->db->exec($sql);
    }

    public function remover($id)
    {
        // BUG: remove livro mesmo se estiver emprestado -> inconsistência de dados
        $sql = "DELETE FROM livros WHERE id = " . $id;
        return $this->db->exec($sql);
    }

    // Retorna a quantidade disponível de um livro
    public function quantidadeDisponivel($id)
    {
        // BUG LÓGICO: considera apenas empréstimos com status 'emprestado',
        // porém não subtrai da quantidade total corretamente se houver devoluções parciais.
        $livro = $this->porId($id);
        if (!$livro) {
            return 0;
        }
        $total = (int)$livro['quantidade'];
        $sql = "SELECT COUNT(*) AS t FROM emprestimos
                WHERE livro_id = $id AND status = 'emprestado'";
        $r = $this->db->query($sql);
        if ($r === false) {
            return $total;
        }
        $emprestados = (int)$r->fetch(PDO::FETCH_ASSOC)['t'];
        return $total - $emprestados;
    }

    // BUG: método "disponibilidade" não usado em lugar algum - código morto.
    public function disponibilidade($id)
    {
        $qtd = $this->quantidadeDisponivel($id);
        if ($qtd <= 0) {
            return 'indisponivel';
        } elseif ($qtd == 1) {
            return 'ultima-unidade';
        }
        return 'disponivel';
    }
}

File: QTS/3° Bim/app-para-teste/desktop/src/main/java/br/edu/etec/biblioteca/models/Livro.java
package br.edu.etec.biblioteca.models;

import br.edu.etec.biblioteca.core.Database;

import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.List;

/**
 * Model Livro.
 * ATENÇÃO: código intencionalmente com FALHAS (SQL injection, regra de negócio).
 */
public class Livro {

    private final Connection conn;

    public Livro() {
        this.conn = Database.getConnection();
    }

    public List<String[]> todos() {
        List<String[]> lista = new ArrayList<>();
        String sql = "SELECT id, titulo, autor, ano, categoria, quantidade, isbn FROM livros";
        try (Statement st = conn.createStatement();
             ResultSet rs = st.executeQuery(sql)) {
            while (rs.next()) {
                lista.add(new String[]{
                    String.valueOf(rs.getInt("id")),
                    rs.getString("titulo"),
                    rs.getString("autor"),
                    String.valueOf(rs.getInt("ano")),
                    rs.getString("categoria"),
                    String.valueOf(rs.getInt("quantidade")),
                    rs.getString("isbn")
                });
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return lista;
    }

    /**
     * Busca livros por termo (título ou autor).
     * BUG DE SEGURANÇA + PERFORMANCE: SQL injection e LIKE sem escape;
     * uso de '%' e '_' do usuário viram curingas (falsos positivos).
     */
    public List<String[]> buscar(String termo) {
        List<String[]> lista = new ArrayList<>();
        String sql = "SELECT id, titulo, autor, ano, categoria, quantidade, isbn FROM livros "
            + "WHERE titulo LIKE '%" + termo + "%' OR autor LIKE '%" + termo + "%'";
        try (Statement st = conn.createStatement();
             ResultSet rs = st.executeQuery(sql)) {
            while (rs.next()) {
                lista.add(new String[]{
                    String.valueOf(rs.getInt("id")),
                    rs.getString("titulo"),
                    rs.getString("autor"),
                    String.valueOf(rs.getInt("ano")),
                    rs.getString("categoria"),
                    String.valueOf(rs.getInt("quantidade")),
                    rs.getString("isbn")
                });
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return lista;
    }

    public String[] porId(int id) {
        String sql = "SELECT id, titulo, autor, ano, categoria, quantidade, isbn FROM livros WHERE id = " + id;
        try (Statement st = conn.createStatement();
             ResultSet rs = st.executeQuery(sql)) {
            if (rs.next()) {
                return new String[]{
                    String.valueOf(rs.getInt("id")),
                    rs.getString("titulo"),
                    rs.getString("autor"),
                    String.valueOf(rs.getInt("ano")),
                    rs.getString("categoria"),
                    String.valueOf(rs.getInt("quantidade")),
                    rs.getString("isbn")
                };
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return null;
    }

    /**
     * Cria livro.
     * BUG DE VALIDAÇÃO: não valida quantidade negativa.
     * BUG DE SEGURANÇA: SQL injection.
     */
    public boolean criar(String titulo, String autor, int ano, String categoria, int quantidade, String isbn) {
        String sql = "INSERT INTO livros (titulo, autor, ano, categoria, quantidade, isbn) VALUES ("
            + "'" + titulo + "', '" + autor + "', " + ano + ", '" + categoria + "', "
            + quantidade + ", '" + isbn + "')";
        try (Statement st = conn.createStatement()) {
            st.executeUpdate(sql);
            return true;
        } catch (SQLException e) {
            return false;
        }
    }

    /**
     * Remove livro.
     * BUG DE INTEGRIDADE: remove mesmo se houver empréstimos ativos.
     */
    public boolean remover(int id) {
        String sql = "DELETE FROM livros WHERE id = " + id;
        try (Statement st = conn.createStatement()) {
            st.executeUpdate(sql);
            return true;
        } catch (SQLException e) {
            return false;
        }
    }

    /**
     * Quantidade disponível de um livro.
     * BUG LÓGICO: conta apenas empréstimos com status 'emprestado' e subtrai da
     * quantidade total. Se houver mais empréstimos ativos que o estoque, retorna
     * valor NEGATIVO (possível porque o emprestar() não checa disponibilidade).
     */
    public int quantidadeDisponivel(int id) {
        String[] livro = porId(id);
        if (livro == null) {
            return 0;
        }
        int total = Integer.parseInt(livro[5]);
        String sql = "SELECT COUNT(*) AS t FROM emprestimos "
            + "WHERE livro_id = " + id + " AND status = 'emprestado'";
        try (Statement st = conn.createStatement();
             ResultSet rs = st.executeQuery(sql)) {
            rs.next();
            int emprestados = rs.getInt("t");
            return total - emprestados;
        } catch (SQLException e) {
            e.printStackTrace();
            return total;
        }
    }

    /**
     * BUG: método morto (código morto) — não é chamado em nenhum lugar da UI.
     */
    public String disponibilidade(int id) {
        int qtd = quantidadeDisponivel(id);
        if (qtd <= 0) {
            return "indisponivel";
        } else if (qtd == 1) {
            return "ultima-unidade";
        }
        return "disponivel";
    }
}
